### ISSUE TITLE
Add some sort of batched reduction to `desc.batching`

### ISSUE THREAD

**Issue by f0uriest** [PlasmaControl/DESC#1564]
not for this PR, but at some point might be worth implementing some sort of batched reduction `desc.batching`

_Originally posted by @f0uriest in https://github.com/PlasmaControl/DESC/pull/1440#discussion_r1940450193_
            

Basically, if we need to evaluate some expensive function over a bunch of inputs and then reduce it (eg sum). Currently the two main options are a single loop which is slow on GPU, or a full vmap which means materializing the full array in memory (or perhaps hoping the the compiler fuses stuff so it isn't but still). One could potentially improve the performance of the loop by unrolling part of it though that increases compile time. Ideally you want something that takes advantage of the fact that the loop iterations are independent.

semantically should be equivalent to the following:

```python
def unbatched_reduce(fun, x, reduction=jnp.add):
    out = 0
    for xi in x:
        out = reduction(out, fun(xi))
    return out

**Comment by unalmis** [comment]
@f0uriest if you think this can be useful partially applying a Jacobian/Hessian before materializing it completely, then please make an issue for that at some point

**Comment by YigitElma** [comment]
> Basically, if we need to evaluate some expensive function over a bunch of inputs and then reduce it (eg sum). Currently the two main options are a single loop which is slow on GPU, or a full vmap which means materializing the full array in memory (or perhaps hoping the the compiler fuses stuff so it isn't but still). One could potentially improve the performance of the loop by unrolling part of it though that increases compile time. Ideally you want something that takes advantage of the fact that the loop iterations are independent.

Maybe `jax.lax.reduce` can be another option

**Comment by unalmis** [comment]
Jax reduce does the reduction in parallel. The point of chunked reduce vmap is to do the reduction in sequences of chunks, so jax.lax.reduce is not useful for chunking.
